### PR TITLE
Allow adjusting a request/response log level

### PR DESCRIPTION
**Pull request by snaiper80**
Related #2250
Motivation:
A client wants to adjust the log level of a request or a response under certain conditions.
For example, a response that has a certain exception may not be a failure.
But it always logs as an error.

Modifications:
- Add request/response log level mapper function in `Logging{Client,Service}`.
- Add the default mapping function to return an original log level for backward compatibility.
- Simplify `logRequest()`, and `logResponse()`

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/logging/LoggingClient.java [hunk 1]
```diff
@@ -113,6 +116,8 @@ public LoggingClient(Client<I, O> delegate, LogLevel level) {
              level,
              level,
              level,
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/logging/LoggingClient.java]: So these three above are not used?

File: core/src/main/java/com/linecorp/armeria/server/logging/LoggingService.java [hunk 2]
```diff
@@ -80,6 +83,8 @@ public static LoggingServiceBuilder builder() {
             LogLevel requestLogLevel,
             LogLevel successfulResponseLogLevel,
             LogLevel failedResponseLogLevel,
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/logging/LoggingService.java]: ditto. These three above are not used.

Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/server/logging/LoggingService.java]: Oops. I forgot it.

File: core/src/main/java/com/linecorp/armeria/client/logging/LoggingClient.java (hunk 1 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/client/logging/LoggingClient.java]: I forgot to remove it. Thank you @minwoox

**Comment by snaiper80** [comment]
@trustin Raising IllegalStateException is also confused in my opinion. A user cannot recognized this concept is conflicted or not on they code. How about change the variable name? 

Just like below...
requestLogLevel() -> defaultRequestLogLevel()
successfulResponseLogLevel() -> defaultSuccessfulResponseLogLevel()
failedResponseLogLevel() -> defaultSuccessfulResponseLogLevel()

I think it is natural because the default mapping function has called this methods(or variables) too.

Is it too dangerous?

**Comment by trustin** [comment]
I think renaming doesn’t help much. The point of the suggestion is that it ‘notifies’ a user of wrong configuration before the server/client starts up silently. Please also note that the original name aligns better in terms of consistency across our API.

**Comment by snaiper80** [comment]
@trustin OK. Consistency of API is important. I'll try it.

**Comment by snaiper80** [comment]
@trustin Added state validation code.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 3]
```diff
@@ -49,7 +58,11 @@
      * Sets the {@link LogLevel} to use when logging requests. If unset, will use {@link LogLevel#TRACE}.
      */
     public T requestLogLevel(LogLevel requestLogLevel) {
+        if (this.isSetRequestLogLevelMapper) {
+            throw new IllegalStateException("Request log level mapper was already set.");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: `requestLogLevelMapper has been set already.`

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 4]
```diff
@@ -65,8 +78,12 @@ protected LogLevel requestLogLevel() {
      * If unset, will use {@link LogLevel#TRACE}.
      */
     public T successfulResponseLogLevel(LogLevel successfulResponseLogLevel) {
+        if (this.isSetResponseLogLevelMapper) {
+            throw new IllegalStateException("Response log level mapper was already set.");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: `responseLogLevelMapper has been set already.`

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 5]
```diff
@@ -82,7 +99,11 @@ protected LogLevel successfulResponseLogLevel() {
      * If unset, will use {@link LogLevel#WARN}. The request will be logged too if it was not otherwise.
      */
     public T failureResponseLogLevel(LogLevel failedResponseLogLevel) {
+        if (this.isSetResponseLogLevelMapper) {
+            throw new IllegalStateException("Response log level mapper was already set.");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: `responseLogLevelMapper has been set already.`

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 6]
```diff
@@ -93,6 +114,44 @@ protected LogLevel failedResponseLogLevel() {
         return failedResponseLogLevel;
     }
 
+    /**
+     * Sets the {@link Function} to use when mapping the log level of request logs.
+     */
+    public T requestLogLevelMapper(Function<? super RequestLog, LogLevel> requestLogLevelMapper) {
+        if (this.isSetRequestLogLevel) {
+            throw new IllegalStateException("Request log level was already set.");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: `requestLogLevel has been set already.`

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 7]
```diff
@@ -93,6 +114,44 @@ protected LogLevel failedResponseLogLevel() {
...
+    protected Function<? super RequestLog, LogLevel> requestLogLevelMapper() {
+        return requestLogLevelMapper;
+    }
+
+    /**
+     * Sets the {@link Function} to use when mapping the log level of response logs.
+     */
+    public T responseLogLevelMapper(Function<? super RequestLog, LogLevel> responseLogLevelMapper) {
+        if (this.isSetSuccessfulResponseLogLevel || this.isSetFailedResponseLogLevel) {
+            throw new IllegalStateException("Response log level was already set.");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Could you use different message for each case?

- `successfulResponseLogLevel has been set already.`
- `failedResponseLogLevel has been set already.`

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 8]
```diff
@@ -281,7 +343,10 @@ public String toString() {
         final ToStringHelper helper = MoreObjects.toStringHelper(self)
                                                  .add("requestLogLevel", requestLogLevel)
                                                  .add("successfulResponseLogLevel", successfulResponseLogLevel)
-                                                 .add("failedResponseLogLevel", failureResponseLogLevel);
+                                                 .add("failedResponseLogLevel", failureResponseLogLevel)
+                                                 .add("requestLogLevelMapper", requestLogLevelMapper)
+                                                 .add("responseLogLevelMapper", responseLogLevelMapper);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: How about adding these properties selectively, like:

```java
if (isSetRequestLogLevelMapper) {
    helper.add("requestLogLevelMapper", requestLogLevelMapper);
} else {
    helper.add("requestLogLevel", requestLogLevel);
}
if (isSetResponseLogLevelMapper) {
    helper.add("responseLogLevelMapper", responseLogLevelMapper);
} else {
    helper.add("successfulResponseLogLevel", successfulResponseLogLevel);
    helper.add("failedResponseLogLevel", failedResponseLogLevel);
}
```

File: core/src/main/java/com/linecorp/armeria/internal/logging/LoggingDecorators.java [hunk 9]
```diff
@@ -56,44 +58,46 @@ public static void logRequest(
...
+
+        if (responseLogLevel.isEnabled(logger)) {
             final String responseStr = log.toStringResponseOnly(responseHeadersSanitizer,
                                                                 responseContentSanitizer,
                                                                 responseTrailersSanitizer);
             if (responseCause == null) {
-                level.log(logger, RESPONSE_FORMAT, responseStr);
+                responseLogLevel.log(logger, RESPONSE_FORMAT, responseStr);
             } else {
                 if (!requestLogLevel.isEnabled(logger)) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/logging/LoggingDecorators.java]: You can delay the evaluation of `requestLogLevelMapper` up to this point.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 10]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
                             ", trailers: " + RESPONSE_TRAILERS);
     }
 
+    @Test
+    public void  map_request_level_test_header_as_warn() throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: - double spaces between `void` and the method name.
- Could you use camel-case notation?

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 11]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+        verify(logger).warn(REQUEST_FORMAT,
+                            "headers: " + REQUEST_HEADERS + ", content: " + REQUEST_CONTENT +
+                            ", trailers: " + REQUEST_TRAILERS);
+        verify(logger).warn(RESPONSE_FORMAT,
+                            "headers: " + RESPONSE_HEADERS + ", content: " + RESPONSE_CONTENT +
+                            ", trailers: " + RESPONSE_TRAILERS);
+    }
+
+    @Test
+    public void map_request_level_header_as_info() throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: - Ditto - camel case
- Could this method be merged with `map_request_level_test_header_as_warn` or could we deduplicate the log mappers?

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 12]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+        verify(logger).info(REQUEST_FORMAT,
+                            "headers: " + REQUEST_HEADERS + ", content: " + REQUEST_CONTENT +
+                            ", trailers: " + REQUEST_TRAILERS);
+        verify(logger).info(RESPONSE_FORMAT,
+                            "headers: " + RESPONSE_HEADERS + ", content: " + RESPONSE_CONTENT +
+                            ", trailers: " + RESPONSE_TRAILERS);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void double_set_request_log_level_and_mapper() throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: Ditto - camel case

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 13]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void double_set_request_log_level_and_mapper_reversed() throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: Ditto - camel case

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 14]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                              .requestLogLevel(LogLevel.INFO)
+                              .requestLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: These two lines could be removed.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 15]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                LoggingService.builder()
+                              .requestLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              }).requestLogLevel(LogLevel.INFO)
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: These two lines could be removed.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 16]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                              .successfulResponseLogLevel(LogLevel.INFO)
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: These two lines could be removed.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 17]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              }).requestLogLevel(LogLevel.INFO)
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void double_set_successful_response_log_level_and_mapper() throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: Ditto - camel case

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 18]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void double_set_successful_response_log_level_and_mapper_reversed() throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: Ditto - camel case

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 19]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
+                              .successfulResponseLogLevel(LogLevel.INFO)
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: These two lines could be removed.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 20]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                              .failureResponseLogLevel(LogLevel.INFO)
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: These two lines could be removed.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 21]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
+                              .failureResponseLogLevel(LogLevel.INFO)
+                              .newDecorator().apply(delegate);
+        service.serve(ctx, REQUEST);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: These two lines could be removed.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 22]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+    public void double_set_failure_response_log_level_and_mapper_reversed() throws Exception {
+        final LoggingService service =
+                LoggingService.builder()
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: `log -> LogLevel.INFO` would be enough.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 23]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+        final LoggingService service =
+                LoggingService.builder()
+                              .failureResponseLogLevel(LogLevel.INFO)
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: `log -> LogLevel.INFO` would be enough.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 24]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+    public void double_set_successful_response_log_level_and_mapper_reversed() throws Exception {
+        final LoggingService service =
+                LoggingService.builder()
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: `log -> LogLevel.INFO` would be enough.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 25]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+        final LoggingService service =
+                LoggingService.builder()
+                              .successfulResponseLogLevel(LogLevel.INFO)
+                              .responseLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: `log -> LogLevel.INFO` would be enough.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 26]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+    @Test(expected = IllegalStateException.class)
+    public void double_set_request_log_level_and_mapper_reversed() throws Exception {
+        final LoggingService service =
+                LoggingService.builder()
+                              .requestLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: `log -> LogLevel.INFO` would be enough.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 27]
```diff
@@ -154,6 +155,165 @@ public void infoLevel() throws Exception {
...
+        final LoggingService service =
+                LoggingService.builder()
+                              .requestLogLevel(LogLevel.INFO)
+                              .requestLogLevelMapper(log -> {
+                                  if (log.requestHeaders().contains("x-test")) {
+                                      return LogLevel.WARN;
+                                  } else {
+                                      return LogLevel.INFO;
+                                  }
+                              })
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: `log -> LogLevel.INFO` would be enough.

**Comment by snaiper80** [comment]
@trustin Changed codes you mentioned. Thank you for your review.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 28]
```diff
@@ -154,6 +155,91 @@ public void infoLevel() throws Exception {
...
+        service.serve(ctx, REQUEST);
+        verify(logger).info(REQUEST_FORMAT,
+                            "headers: " + REQUEST_HEADERS + ", content: " + REQUEST_CONTENT +
+                            ", trailers: " + REQUEST_TRAILERS);
+        verify(logger).info(RESPONSE_FORMAT,
+                            "headers: " + RESPONSE_HEADERS + ", content: " + RESPONSE_CONTENT +
+                            ", trailers: " + RESPONSE_TRAILERS);
+    }
+
+    @Test(expected = IllegalStateException.class)
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: nit: Could you use `assertThatThrownBy()` for an exception matching? We usually prefer asserting with `assertThatXXX` of [`assertj`](https://joel-costigliola.github.io/assertj/assertj-core-quick-start.html) :-)

```java
@Test
public void duplicateSetRequestLogLevelAndMapper() throws Exception {
    assertThatThrownBy(() -> {
        LoggingService.builder()
                      .requestLogLevel(LogLevel.INFO)
                      .requestLogLevelMapper(log -> LogLevel.INFO);
    }).isInstanceOf(IllegalStateException.class);
}
```

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 29]
```diff
@@ -49,7 +58,11 @@
      * Sets the {@link LogLevel} to use when logging requests. If unset, will use {@link LogLevel#TRACE}.
      */
     public T requestLogLevel(LogLevel requestLogLevel) {
+        if (this.isSetRequestLogLevelMapper) {
```
[the remark this answers is not in the document]
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: OK. will fix it.

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java (hunk 28 above)
Comment (snaiper80) [reply core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: Umm..I already checked `HBaseClientCompatibilityTest` case. I think both case are allowed with any style. Anyway will fix it.

Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: Sorry for the confusion. I guess `HBaseClientCompatibilityTest` was made before our current policy. Actually, we now recommend to build test with [JUnit 5](https://github.com/line/armeria/blob/master/CONTRIBUTING.md#use-junit-5-instead-of-junit-4-for-testing). 😉

**Comment by snaiper80** [comment]
@ikhoon Thank you for your kindly feedback.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 30]
```diff
@@ -36,6 +36,15 @@
     private LogLevel requestLogLevel = LogLevel.TRACE;
     private LogLevel successfulResponseLogLevel = LogLevel.TRACE;
     private LogLevel failedResponseLogLevel = LogLevel.WARN;
+    private Function<? super RequestLog, LogLevel> requestLogLevelMapper =
+            log -> requestLogLevel();
+    private Function<? super RequestLog, LogLevel> responseLogLevelMapper =
+            log -> log.responseCause() == null ? successfulResponseLogLevel() : failedResponseLogLevel();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: I think we can make theses three methods as private now.
`requestLogLevel()`, `successfulResponseLogLevel()` and `failedResponseLogLevel()`.
But before that let's make them deprecated first.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 31]
```diff
@@ -36,6 +36,15 @@
...
+    private boolean isSetRequestLogLevel;
+    private boolean isSetSuccessfulResponseLogLevel;
+    private boolean isSetFailedResponseLogLevel;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: I think we can integrate these two flags into one.
`isSetSuccessfulResponseLogLevel`, `isSetFailedResponseLogLevel` ->
`isResponseLogLevelSet`

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 32]
```diff
@@ -36,6 +36,15 @@
     private LogLevel requestLogLevel = LogLevel.TRACE;
     private LogLevel successfulResponseLogLevel = LogLevel.TRACE;
     private LogLevel failedResponseLogLevel = LogLevel.WARN;
+    private Function<? super RequestLog, LogLevel> requestLogLevelMapper =
+            log -> requestLogLevel();
+    private Function<? super RequestLog, LogLevel> responseLogLevelMapper =
+            log -> log.responseCause() == null ? successfulResponseLogLevel() : failedResponseLogLevel();
+    private boolean isSetRequestLogLevel;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: nit: how about `isRequestLogLevelSet`?

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 33]
```diff
@@ -36,6 +36,15 @@
...
+    private boolean isSetSuccessfulResponseLogLevel;
+    private boolean isSetFailedResponseLogLevel;
+    private boolean isSetRequestLogLevelMapper;
+    private boolean isSetResponseLogLevelMapper;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: ditto

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 31 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Looks good. But how do I raise a error message in `responseLogLevelMapper()`. 
In the method, it check 2 conditions about `{Successful|Failed} ResponseLogLevel`.
If the user set the response level only one (e.g success or failed), I don't know what the property is set.  If I'll change this, the error message is not helpful to user I think. Is this good?

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 30 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: @minwoox Do you have a plan to offer the mapper only in the future? I think these methods can set as a default value of the mapper function.

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: I don't think we are going to remove the three setters and they should not be deprecated.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 31 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: If we set `isResponseLogLevelSet` to `true` when either `(Successful|Failed)ResponseLogLevel` is called, we can just check `isResponseLogLevelSet` in `responseLogLevelMapper` and raise an exeption like `throw new IllegalStateException("successfulResponseLogLevel or failedResponseLogLevel has been set already.");`.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 30 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: What he's saying about deprecation are the getters.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 31 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: or just `responseLogLevel has been set already.` would be enough. 😄

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 30 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Exactly, sorry for making you confused. 😉

Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Ah. you mean getter() only right?

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 31 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: I'm fine with either options. If we merge the flags, we could update the exception message, like `successfulResponseLogLevel or failedResponseLogLevel have been set already.`

Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: I think "successfulResponseLogLevel or failedResponseLogLevel has been set already." is more clear to me if I change it.

Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: OK. let me change it.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 32 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: OK. Let me fix it.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 30 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Umm... I agree with @trustin. the deprecation of them is too early decision now.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 31 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Fixed

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 32 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Fixed

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 30 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Pushed the commit about deprecation of three `getters`.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 32 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Anyway. Just curious. IsSetXXX is used in the thrift generated code. What is more proper?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: `is...Set` sounds more natural to me.

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java [hunk 34]
```diff
@@ -108,7 +114,10 @@ public T failureResponseLogLevel(LogLevel failedResponseLogLevel) {
 
     /**
      * Returns the {@link LogLevel} to use when logging failure responses (e.g., failed with an exception).
+     *
+     * @deprecated It will be private in the future.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: private -> removed (ditto for the two other getters)

Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: OK. Changed it. But YOU ALREADY FIXED. 😄

**Comment by snaiper80** [comment]
Thank you for all your feedbacks. And sorry to many defects(?).

File: core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java (hunk 32 above)
Comment (snaiper80) [reply core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java]: Yeah. The naming is too hard to me. 😃

**Pull request merged**